heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs

```diff
@@ -4748,37 +4748,46 @@ mbExBlockToSubShape a =
+-- offset, if possible, assuming the offset is in the block permission. The
+-- supplied function provides a partial substitution from variables of
+-- 'LLVMBlockType' to their shapes, in order to split @eqsh@ shapes.
+splitLLVMBlockPerm ::
+  (1 <= w, KnownNat w) =>
+  (ExprVar (LLVMBlockType w) -> Maybe (PermExpr (LLVMShapeType w))) ->
+  PermExpr (BVType w) -> LLVMBlockPerm w ->
+  Maybe (LLVMBlockPerm w, LLVMBlockPerm w)
+splitLLVMBlockPerm _ off bp
   | bvEq off (llvmBlockOffset bp)
   = Just (bp { llvmBlockLen = bvInt 0, llvmBlockShape = PExpr_EmptyShape },
           bp)
-splitLLVMBlockPerm off bp@(llvmBlockShape -> PExpr_EmptyShape) =
+splitLLVMBlockPerm _ off bp@(llvmBlockShape -> PExpr_EmptyShape) =
   Just (bp { llvmBlockLen = bvSub off (llvmBlockOffset bp) },
         bp { llvmBlockOffset = off,
              llvmBlockLen = bvSub (llvmBlockEndOffset bp) off })
-splitLLVMBlockPerm off bp@(LLVMBlockPerm { llvmBlockShape = sh })
+splitLLVMBlockPerm blsubst off bp@(LLVMBlockPerm { llvmBlockShape = sh })
   | Just sh_len <- llvmShapeLength sh
   , bvLt sh_len (bvSub off (llvmBlockOffset bp)) =
     -- If we are splitting after the end of the natural length of a shape, then
     -- pad out the block permission to its natural length and fall back to the
     -- sequence shape case below
-    splitLLVMBlockPerm off (bp { llvmBlockShape =
-                                   PExpr_SeqShape sh PExpr_EmptyShape })
-splitLLVMBlockPerm _ (llvmBlockShape -> PExpr_Var _) = Nothing
-splitLLVMBlockPerm off bp@(llvmBlockShape ->
-                           PExpr_NamedShape maybe_rw maybe_l nmsh args)
+    splitLLVMBlockPerm blsubst off (bp { llvmBlockShape =
+                                         PExpr_SeqShape sh PExpr_EmptyShape })
+splitLLVMBlockPerm _ _ (llvmBlockShape -> PExpr_Var _) = Nothing
+splitLLVMBlockPerm blsubst off bp@(llvmBlockShape ->
+                                   PExpr_NamedShape maybe_rw maybe_l nmsh args)
   | TrueRepr <- namedShapeCanUnfoldRepr nmsh
   , Just sh' <- unfoldModalizeNamedShape maybe_rw maybe_l nmsh args =
-    splitLLVMBlockPerm off (bp { llvmBlockShape = sh' })
-splitLLVMBlockPerm _ (llvmBlockShape -> PExpr_NamedShape _ _ _ _) = Nothing
-splitLLVMBlockPerm _ (llvmBlockShape -> PExpr_EqShape _ _) = Nothing
-splitLLVMBlockPerm _ (llvmBlockShape -> PExpr_PtrShape _ _ _) = Nothing
-splitLLVMBlockPerm _ (llvmBlockShape -> PExpr_FieldShape _) = Nothing
-splitLLVMBlockPerm off bp@(llvmBlockShape -> PExpr_ArrayShape len stride sh)
+    splitLLVMBlockPerm blsubst off (bp { llvmBlockShape = sh' })
+splitLLVMBlockPerm _ _ (llvmBlockShape -> PExpr_NamedShape _ _ _ _) = Nothing
+splitLLVMBlockPerm blsubst off bp@(llvmBlockShape ->
+                                   PExpr_EqShape _len (PExpr_Var b))
+  -- FIXME: make sure the returned shape fits into len bytes!
+  | Just sh <- blsubst b
+  = splitLLVMBlockPerm blsubst off (bp { llvmBlockShape = sh })
+splitLLVMBlockPerm _ _ (llvmBlockShape -> PExpr_EqShape _ _) = Nothing
```

Comment:
Nit pick: combine the two `llvmBlockShape -> PExpr_EqShape ...` cases:

```suggestion
splitLLVMBlockPerm blsubst off bp@(llvmBlockShape ->
                                   PExpr_EqShape _len (PExpr_Var b))
  -- FIXME: make sure the returned shape fits into len bytes!
  | Just sh <- blsubst b
  = splitLLVMBlockPerm blsubst off (bp { llvmBlockShape = sh })
  | otherwise
  = Nothing
```

Aside from making this slightly more readable, this should be a little easier for GHC's pattern-match coverage checker to handle.

Reply:
I see what you mean, that it's weird to use a view pattern with a record accessor. I'll fix that.

Reply:
Wait, except I guess that's not what you were suggesting. I guess you were suggesting I use an `otherwise` branch of the pattern-match. That's not actually quite correct, because that would only match when the second argument to `PExpr_EqShape` is of the form `PExpr_Var b`, whereas the current version doesn't match on that argument at all in this case. TBF, `PExpr_Var b` is the only `PermExpr` expression that can match at that particular type.

Maybe this is all bikeshedding, because there already is a catch-all case at the end of the function?

Reply:
Ah, quite right. Indeed, you would need to move the match on `PExpr_Var` into a pattern guard to preserve the existing semantics of this function:

```suggestion
splitLLVMBlockPerm blsubst off bp@(llvmBlockShape ->
                                   PExpr_EqShape _len blk)
  -- FIXME: make sure the returned shape fits into len bytes!
  | PExpr_Var b <- blk
  , Just sh <- blsubst b
  = splitLLVMBlockPerm blsubst off (bp { llvmBlockShape = sh })
  | otherwise
  = Nothing
```

I'll let you make the call on whether the old or new code is more readable.